source/common/router/router.cc

```diff
@@ -541,37 +578,85 @@ void Filter::onRequestComplete() {
+    // Don't record a timeout for upstream requests we've already seen headers
+    // for.
+    if (!upstream_request->upstream_headers_) {
+      cluster_->stats().upstream_rq_timeout_.inc();
+      if (upstream_request->upstream_host_) {
+        upstream_request->upstream_host_->stats().rq_timeout_.inc();
+      }
+
+      // If this upstream request already hit a "soft" timeout, then it
+      // already recorded a timeout into outlier detection. Don't do it again.
+      if (!upstream_request->outlier_detection_timeout_recorded_) {
+        updateOutlierDetection(timeout_response_code_, *upstream_request);
+      }
+      upstream_request->resetStream();
 
-    updateOutlierDetection(timeout_response_code_, *upstream_requests_.front().get());
-    upstream_requests_.front()->resetStream();
+      chargeUpstreamAbort(timeout_response_code_, false, *upstream_request);
+    }
   }
 
   onUpstreamTimeoutAbort(StreamInfo::ResponseFlag::UpstreamRequestTimeout);
 }
 
+// Called when the per try timeout is hit but we didn't reset the request
+// (hedge_on_per_try_timeout enabled).
+void Filter::onSoftPerTryTimeout(UpstreamRequest& upstream_request) {
+  // Even though we didn't cancel the request yet we still want to track it
+  // in outlier detection.
+  // TODO(mpuncel) is it weird to have a pretend response code here? we might
+  // get a 200 back from this request later.
+  updateOutlierDetection(timeout_response_code_, upstream_request);
+  upstream_request.outlier_detection_timeout_recorded_ = true;
+
+  if (!downstream_response_started_ && retry_state_) {
+    RetryStatus retry_status =
+        retry_state_->shouldHedgeRetryPerTryTimeout([this]() -> void { doRetry(); });
+
+    if (retry_status == RetryStatus::Yes && setupRetry()) {
+      setupRetry();
+      // Don't increment upstream_host->stats().rq_error_ here, we'll do that
+      // later if 1) we hit global timeout or 2) we get bad response headers
+      // back.
+      upstream_request.retried_ = true;
+    } else if (retry_status == RetryStatus::NoOverflow) {
+      callbacks_->streamInfo().setResponseFlag(StreamInfo::ResponseFlag::UpstreamOverflow);
+    } else if (retry_status == RetryStatus::NoRetryLimitExceeded) {
+      callbacks_->streamInfo().setResponseFlag(
+          StreamInfo::ResponseFlag::UpstreamRetryLimitExceeded);
+    }
+  }
+}
+
 void Filter::onPerTryTimeout(UpstreamRequest& upstream_request) {
+  if (hedging_params_.hedge_on_per_try_timeout_) {
+    onSoftPerTryTimeout(upstream_request);
+    return;
+  }
+
+  cluster_->stats().upstream_rq_per_try_timeout_.inc();
+  if (upstream_request.upstream_host_) {
+    upstream_request.upstream_host_->stats().rq_timeout_.inc();
+  }
+
+  upstream_request.resetStream();
+
   updateOutlierDetection(timeout_response_code_, upstream_request);
 
   if (maybeRetryReset(Http::StreamResetReason::LocalReset, upstream_request)) {
+    upstream_request.removeFromList(upstream_requests_);
```

Comment:
It scares me a little that we are doing list removals in a return case as specific as this. Is there somewhere more central we could do this? I'm not sure what the options would be.